server/client.go

```diff
@@ -301,73 +365,202 @@ func (c *client) readLoop() {
-				err := cp.bw.Flush()
-				cp.nc.SetWriteDeadline(time.Time{})
-				if err != nil {
-					c.Debugf("Error flushing: %v", err)
-					cp.mu.Unlock()
-					cp.closeConnection()
-					cp.mu.Lock()
-				} else {
-					// Update outbound last activity.
-					cp.last = last
-					// Check if we should tune the buffer.
-					sz := cp.bw.Available()
-					// Check for expansion opportunity.
-					if wfc > 2 && sz <= maxBufSize/2 {
-						cp.bw = bufio.NewWriterSize(cp.nc, sz*2)
-					}
-					// Check for shrinking opportunity.
-					if wfc == 0 && sz >= minBufSize*2 {
-						cp.bw = bufio.NewWriterSize(cp.nc, sz/2)
-					}
-				}
+			// Update last activity for message delivery
+			cp.last = last
+			cp.out.fsp--
+			if budget > 0 && cp.flushOutbound() {
+				budget -= cp.out.lft
+			} else {
+				cp.flushSignal()
 			}
 			cp.mu.Unlock()
 			delete(c.pcd, cp)
 		}
-		// Check to see if we got closed, e.g. slow consumer
+
+		// Update activity, check read buffer size.
 		c.mu.Lock()
 		nc := c.nc
+
 		// Activity based on interest changes or data/msgs.
-		if c.cache.inMsgs > 0 || c.cache.subs > 0 {
+		if c.in.msgs > 0 || c.in.subs > 0 {
 			c.last = last
 		}
+
+		if n >= cap(b) {
+			c.in.srs = 0
+		} else if n < cap(b)/2 { // divide by 2 b/c we want less than what we would shrink to.
+			c.in.srs++
+		}
+
+		// Update read buffer size as/if needed.
+		if n >= cap(b) && cap(b) < maxBufSize {
+			// Grow
+			c.in.rsz = cap(b) * 2
+			b = make([]byte, c.in.rsz)
+		} else if n < cap(b) && cap(b) > minBufSize && c.in.srs > shortsToShrink {
+			// Shrink, for now don't accelerate, ping/pong will eventually sort it out.
+			c.in.rsz = cap(b) / 2
+			b = make([]byte, c.in.rsz)
+		}
 		c.mu.Unlock()
+
+		// Check to see if we got closed, e.g. slow consumer
 		if nc == nil {
 			return
 		}
+	}
+}
+
+// collapsePtoNB will place primary onto nb buffer as needed in prep for WriteTo.
+// This will return a copy on purpose.
+func (c *client) collapsePtoNB() net.Buffers {
+	if c.out.p != nil {
+		p := c.out.p
+		c.out.p = nil
+		return append(c.out.nb, p)
+	}
+	return c.out.nb
+}
+
+// This will handle the fixup needed on a partial write.
+// Assume pending has been already calculated correctly.
+func (c *client) handlePartialWrite(pnb net.Buffers) {
+	nb := c.collapsePtoNB()
+	// The partial needs to be first, so append nb to pnb
+	c.out.nb = append(pnb, nb...)
+}
+
+// flushOutbound will flush outbound buffer to a client.
+// Will return if data was attempted to be written.
+// Lock must be held
+func (c *client) flushOutbound() bool {
+	if c.flags.isSet(flushOutbound) {
+		return false
+	}
+	c.flags.set(flushOutbound)
+	defer c.flags.clear(flushOutbound)
+
+	// Check for nothing to do.
+	if c.nc == nil || c.srv == nil || c.out.pb == 0 {
+		return true // true because no need to queue a signal.
+	}
+
+	// Snapshot opts
+	srv := c.srv
+
+	// Place primary on nb, assign primary to secondary, nil out nb and secondary.
```

Comment:
assign secondary to primary

Reply:
I read this as assign foo = 3, which when read assign primary to secondary is what happens.

Reply:
Sorry for my english: I read `foo = 3` as assign 3 to foo, not foo to 3, hence if you have `p = s` I would have say assign secondary to primary.